1_ Entender datos (pdf, href formula, href, Unnamed: 22, Unnamed: 23???)
2_ Comprobar columnas o filas duplicadas.
3_ Comprobar columnas o filas NaN.
4_ Eliminar columnas (pdf, href formula, href, Unnamed: 22, Unnamed: 23); lo que no dé valor.
5_ Cambiar nombres columnas.

In [2]:
# Libraries
import numpy as np
import pandas as pd

Cargamos datos de formato CSV

In [3]:
attacks = pd.read_csv("../data/attacks.csv", encoding = "ISO-8859-1")

Visualizamos las primeras filas

In [4]:
attacks.head(3)

Case Number         Date    Year        Type Country        Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid     USA      Hawaii   

                         Location  Activity             Name Sex   ...  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu   Surfing      John Denges    M  ...   

      Species           Investigator or Source                     pdf  \
0  White shark                R. Collier, GSAF    2018.06.25-Wolfe.pdf   
1          NaN  K.McMurray, TrackingSharks.com  2018.06.18-McNeely.pdf   
2          NaN  K.McMurray, TrackingSharks.com   2018.06.09-Denges.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  

[3 rows x 24 columns]

 Visualizo las últimas filas hola

In [5]:
attacks.tail(3)

Case Number Date  Year Type Country Area Location Activity Name Sex   \
25720         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25721         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722          xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
25720  ...      NaN                    NaN  NaN          NaN  NaN   
25721  ...      NaN                    NaN  NaN          NaN  NaN   
25722  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
25720           NaN           NaN            NaN         NaN         NaN  
25721           NaN           NaN            NaN         NaN         NaN  
25722           NaN           NaN            NaN         NaN         NaN  

[3 rows x 24 columns]

Visualizo muestra aleatoria

In [6]:
attacks.sample(3)

Case Number Date  Year Type Country Area Location Activity Name Sex   \
13448         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
23380         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
22282         NaN  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       ... Species  Investigator or Source  pdf href formula href  \
13448  ...      NaN                    NaN  NaN          NaN  NaN   
23380  ...      NaN                    NaN  NaN          NaN  NaN   
22282  ...      NaN                    NaN  NaN          NaN  NaN   

      Case Number.1 Case Number.2 original order Unnamed: 22 Unnamed: 23  
13448           NaN           NaN            NaN         NaN         NaN  
23380           NaN           NaN            NaN         NaN         NaN  
22282           NaN           NaN            NaN         NaN         NaN  

[3 rows x 24 columns]

ANALISIS EXPLORATORIO DEL DATAFRAME

In [7]:
attacks.shape

(25723, 24)

In [8]:
attacks.info

<bound method DataFrame.info of       Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Gly

In [10]:
attacks.isnull().sum()

Case Number               17021
Date                      19421
Year                      19423
Type                      19425
Country                   19471
Area                      19876
Location                  19961
Activity                  19965
Name                      19631
Sex                       19986
Age                       22252
Injury                    19449
Fatal (Y/N)               19960
Time                      22775
Species                   22259
Investigator or Source    19438
pdf                       19421
href formula              19422
href                      19421
Case Number.1             19421
Case Number.2             19421
original order            19414
Unnamed: 22               25722
Unnamed: 23               25721
dtype: int64

In [12]:
attacks.describe()

Year  original order
count  6300.000000     6309.000000
mean   1927.272381     3155.999683
std     281.116308     1821.396206
min       0.000000        2.000000
25%    1942.000000     1579.000000
50%    1977.000000     3156.000000
75%    2005.000000     4733.000000
max    2018.000000     6310.000000

In [13]:
attacks.describe().T

count         mean          std  min     25%     50%     75%  \
Year            6300.0  1927.272381   281.116308  0.0  1942.0  1977.0  2005.0   
original order  6309.0  3155.999683  1821.396206  2.0  1579.0  3156.0  4733.0   

                   max  
Year            2018.0  
original order  6310.0

COMPRUEBA DATOS NUMËRICOS Y CATEGÓRICOS

In [16]:
# DATOS NUMÉRICOS
numericas = attacks._get_numeric_data()
numericas.head(3)

Year  original order
0  2018.0          6303.0
1  2018.0          6302.0
2  2018.0          6301.0

In [17]:
# DATOS CATEGÓRICOS
attacks.select_dtypes(include="object").head()

Case Number         Date        Type    Country             Area  \
0  2018.06.25  25-Jun-2018     Boating        USA       California   
1  2018.06.18  18-Jun-2018  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

   ...           Time         Species           Investigator or Source  \
0  ...          18h00      White shark                R. Collier, GSAF   
1  ...  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2  ...          07h45              NaN  K.McMurray, TrackingSharks.com   
3  ...            NaN        2 m shark                  B. Myatt, GSAF   
4  ...            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 Unnamed: 22 Unnamed: 23  
0    2018.06.25         NaN         NaN  
1    2018.06.18         NaN         NaN  
2    2018.06.09         NaN         NaN  
3    2018.06.08         NaN         NaN  
4    2018.06.04         NaN         NaN  

[5 rows x 22 columns]

RENOMBRO COLUMNAS

In [68]:
attacks_copia = attacks.copy()

In [69]:
columnas = list(attacks.columns)
columnas

['case_number',
 'date',
 'year',
 'type',
 'country',
 'area',
 'location',
 'activity',
 'name',
 'sex_',
 'age',
 'injury',
 'fatal_(y/n)',
 'time',
 'species_',
 'investigator_or_source',
 'pdf',
 'href_formula',
 'href',
 'case_number.1',
 'case_number.2',
 'original_order',
 'unnamed:_22',
 'unnamed:_23']

In [70]:
diccio_nuevas = {columna: columna.replace(" ", "_").lower() for columna in columnas}

In [71]:
diccio_nuevas

{'case_number': 'case_number',
 'date': 'date',
 'year': 'year',
 'type': 'type',
 'country': 'country',
 'area': 'area',
 'location': 'location',
 'activity': 'activity',
 'name': 'name',
 'sex_': 'sex_',
 'age': 'age',
 'injury': 'injury',
 'fatal_(y/n)': 'fatal_(y/n)',
 'time': 'time',
 'species_': 'species_',
 'investigator_or_source': 'investigator_or_source',
 'pdf': 'pdf',
 'href_formula': 'href_formula',
 'href': 'href',
 'case_number.1': 'case_number.1',
 'case_number.2': 'case_number.2',
 'original_order': 'original_order',
 'unnamed:_22': 'unnamed:_22',
 'unnamed:_23': 'unnamed:_23'}

In [72]:
attacks_renombrado = attacks.rename(columns=diccio_nuevas)
attacks_renombrado.sample()

case_number                  date    year        type    country  \
5315  1916.04.25.R  Reported 25-Apr-1916  1916.0  Unprovoked  AUSTRALIA   

                 area      location  activity               name sex_  ...  \
5315  New South Wales  Manley Beach  Swimming  Thomas Harrington    M  ...   

     species_           investigator_or_source                          pdf  \
5315      NaN  Modesto Evening News, 4/25/1916  1916.04.25.R-Harrington.pdf   

                                           href_formula  \
5315  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
5315  http://sharkattackfile.net/spreadsheets/pdf_di...  1916.04.25.R   

     case_number.2 original_order unnamed:_22 unnamed:_23  
5315  1916.04.25.R          988.0         NaN         NaN  

[1 rows x 24 columns]

In [73]:
attacks.rename(columns=diccio_nuevas, inplace=True)

In [74]:
attacks.head(2)

case_number         date    year        type country        area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating     USA  California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked     USA     Georgia   

                         location  activity             name sex_  ...  \
0     Oceanside, San Diego County  Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County  Standing  Adyson McNeely     F  ...   

      species_          investigator_or_source                     pdf  \
0  White shark                R. Collier, GSAF    2018.06.25-Wolfe.pdf   
1          NaN  K.McMurray, TrackingSharks.com  2018.06.18-McNeely.pdf   

                                        href_formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href case_number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   

  case_number.2 original_order unnamed:_22 unnamed:_23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  

[2 rows x 24 columns]

EXPLORA DATOS POR COLUMNAS (apuntes)

Columna pdf hay 19421 NaN

In [65]:
pdf = list(attacks["pdf"])

In [66]:
n = 0
for i in pdf:
    if 'pdf' not in i:
        n = n + 1
print(n)

TypeError: argument of type 'float' is not iterable

In [36]:
attacks["href_formula"]

0        http://sharkattackfile.net/spreadsheets/pdf_di...
1        http://sharkattackfile.net/spreadsheets/pdf_di...
2        http://sharkattackfile.net/spreadsheets/pdf_di...
3        http://sharkattackfile.net/spreadsheets/pdf_di...
4        http://sharkattackfile.net/spreadsheets/pdf_di...
                               ...                        
25718                                                  NaN
25719                                                  NaN
25720                                                  NaN
25721                                                  NaN
25722                                                  NaN
Name: href_formula, Length: 25723, dtype: object

In [82]:
attacks[attacks["unnamed:_23"].notna()]

case_number                  date    year        type  \
4415    1952.03.30           30-Mar-1952  1952.0  Unprovoked   
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0    Provoked   

                   country         area                    location  \
4415  NETHERLANDS ANTILLES      Curacao                         NaN   
5840                   USA  Connecticut  Branford, New Haven County   

                                             activity              name sex_  \
4415  Went to aid of child being menaced by the shark       A.J. Eggink    M   
5840                                          Fishing  Captain Pattison    M   

      ...                                           species_  \
4415  ...  Bull shark, 2.7 m [9'] was captured & dragged ...   
5840  ...                                                NaN   

                                 investigator_or_source  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   
5840                       St. Joseph Herald, 9/14/1878   

                            pdf  \
4415      1952.03.30-Eggink.pdf   
5840  1878.09.14.R-Pattison.pdf   

                                           href_formula  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href case_number.1  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...  1878.09.14.R   

     case_number.2 original_order unnamed:_22      unnamed:_23  
4415    1952.03.30         1888.0         NaN           Teramo  
5840  1878.09.14.R          463.0         NaN  change filename  

[2 rows x 24 columns]

OPERA CON COLUMNAS (SELECCIONA, ELIMINA, Valores únicos ...)

Filtramos los datos por el valor de las columnas

In [33]:
attacks.area.unique()

array(['California', 'Georgia', 'Hawaii', 'New South Wales', 'Colima',
       'Pernambuco', 'Florida', 'Queensland', 'South Carolina',
       'Cornwall', 'Westerm Australia', 'Eastern Cape Province',
       'Hua Hin', 'Cocos Island', 'Western Australia',
       'Alifu Alifu Atoll', 'Western Cape Province', 'New Providence',
       'Alagoas', nan, 'Victoria', 'KwaZulu-Natal',
       'Boi Island, Victoria', 'Galapagos Islands', 'Fernando de Noronha',
       'Sepang', 'Holquin Province', 'Pamplemousses ', 'South Australia',
       'North Island', 'New York', 'Canary Islands', ' Upolu Island',
       'Shizuoka Prefecture', 'Texas', 'Castellón', 'Massachusetts',
       'Red Sea Protectorate', 'New Providence District',
       '40 miles off Grand Bahama Island', 'Ascension Island',
       'New Jersey', 'Majorca', 'Washington', 'Tabasco', 'Anjouan',
       'Ibiza Island', 'Marquesas', 'South Devon', 'New Providence ',
       'Sharjah, ', 'Baja California Sur', 'Saint-Leu', 'South Island',
   

BORRAR FILAS Y COLUMNAS. Reset_index y set_index

APPLY!!

Df .loc e .iloc
Seleccionar filas y columnas en Pandas mediante posición con iloc
Seleccionar filas y columnas en Pandas en base a etiquetas con loc

Ordenamos un dataframe
Buscamos valores nulos
Borramos los valores nulos
rellenar los datos que faltan
 Borramos los duplicados

 Agregación de datos
 GroupBy

Decide a HYPOTHESIS (or hypotheses) with which you will clean the data

..............

EXPLORE the DATA and write down what you have found
you can use: df.describe(), df["column"], etc.

DRAW at least 2 GRAPHS that are insightful.

Use at least 5 data cleaning techniques inside a file named clean.ipynb
null values, columns drop, duplicated data, string manipulation, apply fn, categorize, regex, etc.

Show data that validates the conclusions based on your hypoteses in a file named analysis.ipynb

Build a compelling story-telling around your findings. Think of your stakeholders and convince them with your conclusions! (Some slides with few text and pretty plots are normally useful)